### PR TITLE
EQL: Sequence/Join parsing and model

### PR DESCRIPTION
**Pull request by costin**
PR that parses the grammar for EQL join and sequence and creates their respective domain model.
The physical plan and execution will follow-up up in a separate PR.

### CODE REVIEW

**Comment by costin** [comment]
Considering the grammar around timeunits and fork is still discussed, I've let the parser and is and did not spend additional time on the tests; once the that part of the grammar gets locked again, this needs to be revisited.
However the existing test are passing with minor modifications.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -44,11 +61,157 @@ public LogicalPlan visitEventQuery(EqlBaseParser.EventQueryContext ctx) {
...
+                    case "hours":
+                        timeUnit = TimeUnit.HOURS;
+                        break;
+                    case "d":
+                    case "ds":
+                    case "day":
+                    case "days":
+                        timeUnit = TimeUnit.DAYS;
+                        break;
+                    default:
```
Comment (rw-access) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:206]: I added #54760 to track an issue for removing unknown time units and only supporting ones that are recognized by range queries.

**Comment by costin** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by costin** [comment]
@elasticmachine update branch

**Comment by costin** [comment]
@elasticmachine update branch

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -44,11 +61,157 @@ public LogicalPlan visitEventQuery(EqlBaseParser.EventQueryContext ctx) {
...
+        if (ctx == null) {
+            return TimeValue.MINUS_ONE;
+        }
+
+        NumberContext numberCtx = ctx.timeUnit().number();
+        if (numberCtx instanceof IntegerLiteralContext) {
+            Number number = (Number) visitIntegerLiteral((IntegerLiteralContext) numberCtx).fold();
+            long value = number.longValue();
+            
+            if (value <= 0) {
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:171]: does the sequence of 0 timespan make sense at all?

**Review by aleksmaus — APPROVED** [review]

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:171]: I think not - if no maxspan is needed I opted for not declaring it.
If we agree to have `0` indicate no limit across EQL, the parser can be improved.

**Pull request merged**